File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func createBinaryTree(descriptions [][]int) *TreeNode {
    nodes := map[int]*TreeNode{} // set nodes (1 on 1)
    isChilds := map[int]bool{}
    for _, d := range descriptions {
        parent, child, isLeft := d[0], d[1], d[2]

        if nodes[parent] == nil {
            nodes[parent] = &TreeNode{Val: parent}
        }

        if nodes[child] == nil {
            nodes[child] = &TreeNode{Val: child}
        }

        if isLeft == 1 {
            nodes[parent].Left = nodes[child]
        } else {
            nodes[parent].Right = nodes[child]
        }

        isChilds[child] = true
    }

    for val, node := range nodes {
        if !isChilds[val] {
            return node
        }
    }

    return nil
}

// binary tree, hash map, tree
// time: O(n)
// space: O(n)

// note: a node is root if it hasn't became a child

File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        nodes = {}
        is_child = {}
        for parent, child, is_left in descriptions:
            if parent not in nodes:
                nodes[parent] = TreeNode(parent)
            
            if child not in nodes:
                nodes[child] = TreeNode(child)
            
            if is_left:
                nodes[parent].left = nodes[child]
            else:
                nodes[parent].right = nodes[child]
            
            is_child[child] = True
        
        for node in nodes:
            if node not in is_child:
                return nodes[node]
        
        return None

# binary tree, hash map, tree
# time: O(n)
# space: O(n)